<a href="https://colab.research.google.com/github/evangrinalds/DS-Unit-4-Sprint-2-Neural-Networks/blob/main/module4-Deploy/Copy_of_LS_DS18_424_Deploy_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

# Train Practice

## *Data Science Unit 4 Sprint 2 Assignment 4*

Continue to use TensorFlow Keras & a sample of the [Quickdraw dataset](https://github.com/googlecreativelab/quickdraw-dataset) to build a sketch classification model. The dataset has been sampled to only 10 classes and 10000 observations per class. Apply regularization techniques to your model. 

*Don't forgot to switch to GPU on Colab!*

#### Write a function to load your data

In [15]:
!wget https://github.com/LambdaSchool/DS-Unit-4-Sprint-2-Neural-Networks/raw/main/quickdraw10.npz

--2020-10-08 21:16:01--  https://github.com/LambdaSchool/DS-Unit-4-Sprint-2-Neural-Networks/raw/main/quickdraw10.npz
Resolving github.com (github.com)... 192.30.255.112
Connecting to github.com (github.com)|192.30.255.112|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/LambdaSchool/DS-Unit-4-Sprint-2-Neural-Networks/main/quickdraw10.npz [following]
--2020-10-08 21:16:01--  https://raw.githubusercontent.com/LambdaSchool/DS-Unit-4-Sprint-2-Neural-Networks/main/quickdraw10.npz
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25421363 (24M) [application/octet-stream]
Saving to: ‘quickdraw10.npz.1’

quickdraw10.npz.1   100%[===================>]  24.24M  65.1MB/s    in 0.4s    

2020-10-08 21:16:02 (65.1 

In [16]:
from sklearn.model_selection import train_test_split
import numpy as np
def load_quickdraw10(path):

  data = np.load(path)
  X = data['arr_0']
  y = data['arr_1']

  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.3,
                                                      random_state=42)
  
  return X_train, X_test, y_train, y_test

In [17]:
data = np.load('/content/quickdraw10.npz')

In [18]:
X_train, X_test, y_train, y_test = load_quickdraw10('/content/quickdraw10.npz')

In [19]:
X_train.shape
y_train.shape

(70000,)

In [20]:
# Imports
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import SGD, Adam

def create_model(lr=.01, optimizer=SGD):

  opt = optimizer(learning_rate=lr)

  model = Sequential(
      [
      Dense(32, activation='sigmoid', input_dim=784),
      Dense(32, activation='sigmoid'),
      Dense(10, activation='softmax')
      ]
  )

  model.compile(optimizer=opt, loss='sparse_categorical_crossentropy', metrics=['accuracy'])

  return model

#### Batch Size

In [21]:
batch_40 = create_model().fit(X_train, y_train,
                              epochs=15,
                              batch_size=40,
                              validation_data=(X_test, y_test))

Epoch 1/15
1750/1750 [==============================] - 3s 2ms/step - loss: 1.9987 - accuracy: 0.4214 - val_loss: 1.7213 - val_accuracy: 0.5353
Epoch 2/15
1750/1750 [==============================] - 3s 2ms/step - loss: 1.5359 - accuracy: 0.5791 - val_loss: 1.3917 - val_accuracy: 0.5962
Epoch 3/15
1750/1750 [==============================] - 3s 2ms/step - loss: 1.2974 - accuracy: 0.6329 - val_loss: 1.2170 - val_accuracy: 0.6477
Epoch 4/15
1750/1750 [==============================] - 3s 2ms/step - loss: 1.1524 - accuracy: 0.6711 - val_loss: 1.1020 - val_accuracy: 0.6809
Epoch 5/15
1750/1750 [==============================] - 3s 2ms/step - loss: 1.0605 - accuracy: 0.6920 - val_loss: 1.0423 - val_accuracy: 0.6976
Epoch 6/15
1750/1750 [==============================] - 3s 2ms/step - loss: 1.0055 - accuracy: 0.7036 - val_loss: 0.9775 - val_accuracy: 0.7080
Epoch 7/15
1750/1750 [==============================] - 3s 2ms/step - loss: 0.9672 - accuracy: 0.7100 - val_loss: 0.9622 - val_accuracy:

#### Optimizer

In [22]:
from tensorflow.keras.optimizers import RMSprop

rmsprop = create_model(optimizer=RMSprop).fit(X_train, y_train, 
                                          epochs=15, 
                                          batch_size=75, 
                                          validation_data=(X_test, y_test))

Epoch 1/15
934/934 [==============================] - 2s 3ms/step - loss: 1.3186 - accuracy: 0.5473 - val_loss: 1.1177 - val_accuracy: 0.6383
Epoch 2/15
934/934 [==============================] - 2s 3ms/step - loss: 1.0265 - accuracy: 0.6654 - val_loss: 0.9569 - val_accuracy: 0.6919
Epoch 3/15
934/934 [==============================] - 2s 2ms/step - loss: 0.9272 - accuracy: 0.7033 - val_loss: 0.9099 - val_accuracy: 0.7098
Epoch 4/15
934/934 [==============================] - 2s 2ms/step - loss: 0.9107 - accuracy: 0.7113 - val_loss: 0.8684 - val_accuracy: 0.7233
Epoch 5/15
934/934 [==============================] - 2s 2ms/step - loss: 0.8798 - accuracy: 0.7194 - val_loss: 0.8862 - val_accuracy: 0.7135
Epoch 6/15
934/934 [==============================] - 2s 2ms/step - loss: 0.8564 - accuracy: 0.7300 - val_loss: 0.8726 - val_accuracy: 0.7244
Epoch 7/15
934/934 [==============================] - 2s 2ms/step - loss: 0.8532 - accuracy: 0.7312 - val_loss: 0.8676 - val_accuracy: 0.7234
Epoch 

#### Learning Rate

In [24]:
lr_onethousandth = create_model(lr=.001).fit(X_train, y_train,
                                             epochs=15,
                                             batch_size=75,
                                             validation_data=(X_test, y_test))

Epoch 1/15
934/934 [==============================] - 2s 2ms/step - loss: 2.3710 - accuracy: 0.0993 - val_loss: 2.2919 - val_accuracy: 0.1313
Epoch 2/15
934/934 [==============================] - 2s 2ms/step - loss: 2.2544 - accuracy: 0.1988 - val_loss: 2.2235 - val_accuracy: 0.2579
Epoch 3/15
934/934 [==============================] - 2s 2ms/step - loss: 2.2014 - accuracy: 0.2997 - val_loss: 2.1808 - val_accuracy: 0.3389
Epoch 4/15
934/934 [==============================] - 2s 2ms/step - loss: 2.1613 - accuracy: 0.3735 - val_loss: 2.1436 - val_accuracy: 0.3992
Epoch 5/15
934/934 [==============================] - 2s 2ms/step - loss: 2.1256 - accuracy: 0.4232 - val_loss: 2.1098 - val_accuracy: 0.4367
Epoch 6/15
934/934 [==============================] - 2s 2ms/step - loss: 2.0916 - accuracy: 0.4519 - val_loss: 2.0758 - val_accuracy: 0.4616
Epoch 7/15
934/934 [==============================] - 2s 2ms/step - loss: 2.0580 - accuracy: 0.4736 - val_loss: 2.0435 - val_accuracy: 0.4795
Epoch 

#### Regularization

Using your best performing model from the previous module, apply each of the following regularization strategies: 
* Early Stopping
* Dropout
* Weight Decay
* Weight Constraint


#### Early Stopping

In [25]:
from tensorflow.keras.callbacks import EarlyStopping, TensorBoard
from tensorflow.keras.layers import Flatten, Dense
from tensorflow.keras.layers import ReLU
import tensorflow as tf
import os

logdir = os.path.join("logs", "EarlyStopping-Loss")

tensorboard_callback = tf.keras.callbacks.TensorBoard(logdir, histogram_freq=1)
stop = EarlyStopping(monitor='val_loss', min_delta=0.001, patience=5)

model = tf.keras.Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128),
    ReLU(negative_slope=.01),
    Dense(128),
    ReLU(negative_slope=.01),
    Dense(128),
    ReLU(negative_slope=.01),
    Dense(10, activation='softmax')
])

model.compile(loss='sparse_categorical_crossentropy', optimizer='nadam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=99, 
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

Epoch 1/99
   1/2188 [..............................] - ETA: 0s - loss: 88.7144 - accuracy: 0.0312WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/summary_ops_v2.py:1277: stop (from tensorflow.python.eager.profiler) is deprecated and will be removed after 2020-07-01.
Instructions for updating:
use `tf.profiler.experimental.stop` instead.
2188/2188 [==============================] - 11s 5ms/step - loss: 1.5900 - accuracy: 0.6826 - val_loss: 0.7793 - val_accuracy: 0.7635
Epoch 2/99
2188/2188 [==============================] - 11s 5ms/step - loss: 0.6951 - accuracy: 0.7884 - val_loss: 0.6747 - val_accuracy: 0.7970
Epoch 3/99
2188/2188 [==============================] - 11s 5ms/step - loss: 0.6111 - accuracy: 0.8132 - val_loss: 0.6519 - val_accuracy: 0.8032
Epoch 4/99
2188/2188 [==============================] - 11s 5ms/step - loss: 0.5611 - accuracy: 0.8298 - val_loss: 0.5791 - val_accuracy: 0.8257
Epoch 5/99
2188/2188 [==============================] -

In [26]:
%tensorboard --logdir logs

UsageError: Line magic function `%tensorboard` not found.


#### Dropout

In [28]:
from tensorflow.keras.layers import Dropout

from tensorflow.keras.constraints import MaxNorm

wc = MaxNorm(max_value=2)

logdir = os.path.join("logs", "EarlyStopping+WeightConstraint+Dropout")

tensorboard_callback = tf.keras.callbacks.TensorBoard(logdir, histogram_freq=1)
stop = EarlyStopping(monitor='val_loss', min_delta=0.001, patience=3)

model = tf.keras.Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(512, kernel_constraint=wc),
    ReLU(negative_slope=.01),
    Dropout(.2),
    Dense(512, kernel_constraint=wc),
    ReLU(negative_slope=.01),
    Dropout(.2),
    Dense(512, kernel_constraint=wc),
    ReLU(negative_slope=.01),
    Dropout(.2),
    Dense(10, activation='softmax')
])

model.compile(loss='sparse_categorical_crossentropy', optimizer='nadam',
              metrics=['accuracy'])

model.fit(X_train, y_train, epochs=99, 
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

Epoch 1/99
2188/2188 [==============================] - 53s 24ms/step - loss: 2.0093 - accuracy: 0.6748 - val_loss: 0.6757 - val_accuracy: 0.7958
Epoch 2/99
2188/2188 [==============================] - 57s 26ms/step - loss: 0.7383 - accuracy: 0.7803 - val_loss: 0.6112 - val_accuracy: 0.8192
Epoch 3/99
2188/2188 [==============================] - 53s 24ms/step - loss: 0.6768 - accuracy: 0.8012 - val_loss: 0.5640 - val_accuracy: 0.8362
Epoch 4/99
2188/2188 [==============================] - 53s 24ms/step - loss: 0.6456 - accuracy: 0.8112 - val_loss: 0.5626 - val_accuracy: 0.8362
Epoch 5/99
2188/2188 [==============================] - 53s 24ms/step - loss: 0.6261 - accuracy: 0.8192 - val_loss: 0.5539 - val_accuracy: 0.8368
Epoch 6/99
2188/2188 [==============================] - 53s 24ms/step - loss: 0.6212 - accuracy: 0.8231 - val_loss: 0.5631 - val_accuracy: 0.8345
Epoch 7/99
2188/2188 [==============================] - 53s 24ms/step - loss: 0.5983 - accuracy: 0.8269 - val_loss: 0.5704 -

#### Weight Decay aka L1/L2 regularization

In [27]:
from tensorflow.keras import regularizers

logdir = os.path.join("logs", "EarlyStopping+L2_WeightDecay")

tensorboard_callback = tf.keras.callbacks.TensorBoard(logdir, histogram_freq=1)
stop = EarlyStopping(monitor='val_loss', min_delta=0.001, patience=5)

model = tf.keras.Sequential([
    Flatten(input_shape=(28,28)),
    Dense(256, kernel_regularizer=regularizers.l2(0.01)),
    ReLU(negative_slope=.01),
    Dense(256, kernel_regularizer=regularizers.l2(0.01)),
    ReLU(negative_slope=.01),
    Dense(256, kernel_regularizer=regularizers.l2(0.01)),
    ReLU(negative_slope=.01),
    Dense(10, activation='softmax')
])

model.compile(loss='sparse_categorical_crossentropy', optimizer='nadam',
              metrics=['accuracy'])

model.fit(X_train, y_train, epochs=99, 
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

Epoch 1/99
2188/2188 [==============================] - 22s 10ms/step - loss: 6.2260 - accuracy: 0.7350 - val_loss: 2.6148 - val_accuracy: 0.7993
Epoch 2/99
2188/2188 [==============================] - 21s 10ms/step - loss: 1.5061 - accuracy: 0.8174 - val_loss: 0.9629 - val_accuracy: 0.8194
Epoch 3/99
2188/2188 [==============================] - 21s 10ms/step - loss: 0.8680 - accuracy: 0.8191 - val_loss: 0.8720 - val_accuracy: 0.8015
Epoch 4/99
2188/2188 [==============================] - 21s 10ms/step - loss: 0.8056 - accuracy: 0.8197 - val_loss: 0.8152 - val_accuracy: 0.8104
Epoch 5/99
2188/2188 [==============================] - 21s 10ms/step - loss: 0.7905 - accuracy: 0.8221 - val_loss: 0.8531 - val_accuracy: 0.7997
Epoch 6/99
2188/2188 [==============================] - 22s 10ms/step - loss: 0.7937 - accuracy: 0.8250 - val_loss: 0.8200 - val_accuracy: 0.8209
Epoch 7/99
2188/2188 [==============================] - 21s 10ms/step - loss: 0.7746 - accuracy: 0.8260 - val_loss: 0.8383 -

#### Weight Constraint

In [29]:
from tensorflow.keras.constraints import MaxNorm

wc = MaxNorm(max_value=2)

stop = EarlyStopping(monitor='val_loss', min_delta=0.001, patience=3)

model = tf.keras.Sequential([
    Flatten(input_shape=(28,28)),
    Dense(256, kernel_constraint=wc),
    ReLU(negative_slope=.01),
    Dense(256, kernel_constraint=wc),
    ReLU(negative_slope=.01),
    Dense(256, kernel_constraint=wc),
    ReLU(negative_slope=.01),
    Dense(10, activation='softmax')
])

model.compile(loss='sparse_categorical_crossentropy', optimizer='nadam',
              metrics=['accuracy'])

model.fit(X_train, y_train, epochs=99,
          validation_data=(X_test, y_test),
          callbacks=[stop])

Epoch 1/99
2188/2188 [==============================] - 24s 11ms/step - loss: 1.5789 - accuracy: 0.7250 - val_loss: 0.7129 - val_accuracy: 0.7825
Epoch 2/99
2188/2188 [==============================] - 21s 10ms/step - loss: 0.6070 - accuracy: 0.8161 - val_loss: 0.6233 - val_accuracy: 0.8109
Epoch 3/99
2188/2188 [==============================] - 20s 9ms/step - loss: 0.5370 - accuracy: 0.8375 - val_loss: 0.5595 - val_accuracy: 0.8328
Epoch 4/99
2188/2188 [==============================] - 20s 9ms/step - loss: 0.4970 - accuracy: 0.8514 - val_loss: 0.5172 - val_accuracy: 0.8443
Epoch 5/99
2188/2188 [==============================] - 20s 9ms/step - loss: 0.4614 - accuracy: 0.8621 - val_loss: 0.5356 - val_accuracy: 0.8483
Epoch 6/99
2188/2188 [==============================] - 19s 9ms/step - loss: 0.4333 - accuracy: 0.8695 - val_loss: 0.5379 - val_accuracy: 0.8417
Epoch 7/99
2188/2188 [==============================] - 19s 9ms/step - loss: 0.4100 - accuracy: 0.8767 - val_loss: 0.5108 - val_

#### Model Summary

In [30]:
model.summary()

Model: "sequential_10"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_3 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_33 (Dense)             (None, 256)               200960    
_________________________________________________________________
re_lu_9 (ReLU)               (None, 256)               0         
_________________________________________________________________
dense_34 (Dense)             (None, 256)               65792     
_________________________________________________________________
re_lu_10 (ReLU)              (None, 256)               0         
_________________________________________________________________
dense_35 (Dense)             (None, 256)               65792     
_________________________________________________________________
re_lu_11 (ReLU)              (None, 256)             

## Deploy

Save your model's weights using the Checkpoint function. Try reloading the model and making inference on your validation dataset.

In [31]:
import tensorflow as tf

# Save the latest (best) weights of your model at the end of each epoch
cpoint = tf.keras.callbacks.ModelCheckpoint("weights_best.h5",
                                            verbose=1,
                                            save_weights_only=True)
stop = EarlyStopping(monitor='val_loss', min_delta=0.001, patience=3)

def create_model():
  model = tf.keras.Sequential([
      Flatten(input_shape=(28, 28)),
      Dense(128),
      ReLU(negative_slope=.01),
      Dense(128),
      ReLU(negative_slope=.01),
      Dense(128),
      ReLU(negative_slope=.01),
      Dense(10, activation='softmax')
  ])

  model.compile(loss='sparse_categorical_crossentropy', optimizer='nadam',
                metrics=['accuracy'])

  return model


model = create_model()
model.fit(X_train, y_train, epochs=10,
          validation_data=(X_test, y_test),
          verbose=2,
          callbacks=[cpoint, stop])

Epoch 1/10

Epoch 00001: saving model to weights_best.h5
2188/2188 - 9s - loss: 1.6751 - accuracy: 0.6937 - val_loss: 0.7868 - val_accuracy: 0.7580
Epoch 2/10

Epoch 00002: saving model to weights_best.h5
2188/2188 - 9s - loss: 0.6623 - accuracy: 0.7983 - val_loss: 0.6478 - val_accuracy: 0.8000
Epoch 3/10

Epoch 00003: saving model to weights_best.h5
2188/2188 - 9s - loss: 0.5844 - accuracy: 0.8237 - val_loss: 0.5913 - val_accuracy: 0.8234
Epoch 4/10

Epoch 00004: saving model to weights_best.h5
2188/2188 - 9s - loss: 0.5331 - accuracy: 0.8378 - val_loss: 0.5464 - val_accuracy: 0.8357
Epoch 5/10

Epoch 00005: saving model to weights_best.h5
2188/2188 - 9s - loss: 0.5001 - accuracy: 0.8486 - val_loss: 0.5391 - val_accuracy: 0.8377
Epoch 6/10

Epoch 00006: saving model to weights_best.h5
2188/2188 - 9s - loss: 0.4631 - accuracy: 0.8607 - val_loss: 0.5116 - val_accuracy: 0.8482
Epoch 7/10

Epoch 00007: saving model to weights_best.h5
2188/2188 - 9s - loss: 0.4362 - accuracy: 0.8690 - val_

In [32]:
model.summary()

Model: "sequential_11"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_4 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_37 (Dense)             (None, 128)               100480    
_________________________________________________________________
re_lu_12 (ReLU)              (None, 128)               0         
_________________________________________________________________
dense_38 (Dense)             (None, 128)               16512     
_________________________________________________________________
re_lu_13 (ReLU)              (None, 128)               0         
_________________________________________________________________
dense_39 (Dense)             (None, 128)               16512     
_________________________________________________________________
re_lu_14 (ReLU)              (None, 128)             

In [33]:
model.evaluate(X_test, y_test)  # Model val metrics after full training

938/938 [==============================] - 1s 1ms/step - loss: 0.5226 - accuracy: 0.8450


[0.5226046442985535, 0.8450333476066589]

#### Save / Export your Model

In [35]:
model = create_model()  # Recreate basic model architecture (and compiling)
# But we've not trained it, and that's the expensive step
# Instead let's just load the best weights from last time
model.load_weights('weights_best.h5')
model.summary()

Model: "sequential_12"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_5 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_41 (Dense)             (None, 128)               100480    
_________________________________________________________________
re_lu_15 (ReLU)              (None, 128)               0         
_________________________________________________________________
dense_42 (Dense)             (None, 128)               16512     
_________________________________________________________________
re_lu_16 (ReLU)              (None, 128)               0         
_________________________________________________________________
dense_43 (Dense)             (None, 128)               16512     
_________________________________________________________________
re_lu_17 (ReLU)              (None, 128)             

In [36]:
model.evaluate(X_test, y_test)  # Model val metrics after full training

938/938 [==============================] - 1s 1ms/step - loss: 0.5226 - accuracy: 0.8450


[0.5226046442985535, 0.8450333476066589]

In [37]:
model.predict

<bound method Model.predict of <tensorflow.python.keras.engine.sequential.Sequential object at 0x7f431b8da5f8>>

In [38]:
# We can also save the entire model, not just weights
# That way we don't need to create_model() function
# Note - this saves architecture, but seems to lose some .compile details
# i.e. you might still need to recompile to know which metrics to use, etc.

# Create and train a new model instance.
model = create_model()
model.fit(X_train, y_train, epochs=5)

# Save the entire model as a SavedModel. *Not* a callback, just save at the end.
!mkdir -p saved_model
model.save('saved_model/my_model')

Epoch 1/5
2188/2188 [==============================] - 9s 4ms/step - loss: 1.7216 - accuracy: 0.6839
Epoch 2/5
2188/2188 [==============================] - 9s 4ms/step - loss: 0.6988 - accuracy: 0.7861
Epoch 3/5
2188/2188 [==============================] - 9s 4ms/step - loss: 0.6095 - accuracy: 0.8145
Epoch 4/5
2188/2188 [==============================] - 10s 5ms/step - loss: 0.5572 - accuracy: 0.8304
Epoch 5/5
2188/2188 [==============================] - 9s 4ms/step - loss: 0.5200 - accuracy: 0.8432
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
INFO:tensorflow:Assets written to: saved_model/my_model/assets


In [39]:
model.summary()  # Model as trained

Model: "sequential_13"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_6 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_45 (Dense)             (None, 128)               100480    
_________________________________________________________________
re_lu_18 (ReLU)              (None, 128)               0         
_________________________________________________________________
dense_46 (Dense)             (None, 128)               16512     
_________________________________________________________________
re_lu_19 (ReLU)              (None, 128)               0         
_________________________________________________________________
dense_47 (Dense)             (None, 128)               16512     
_________________________________________________________________
re_lu_20 (ReLU)              (None, 128)             

In [40]:
new_model = tf.keras.models.load_model('saved_model/my_model')

# Check its architecture
new_model.summary()

Model: "sequential_13"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_6 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_45 (Dense)             (None, 128)               100480    
_________________________________________________________________
re_lu_18 (ReLU)              (None, 128)               0         
_________________________________________________________________
dense_46 (Dense)             (None, 128)               16512     
_________________________________________________________________
re_lu_19 (ReLU)              (None, 128)               0         
_________________________________________________________________
dense_47 (Dense)             (None, 128)               16512     
_________________________________________________________________
re_lu_20 (ReLU)              (None, 128)             

In [41]:
model.evaluate(X_test, y_test)  # Trained model

938/938 [==============================] - 1s 1ms/step - loss: 0.5914 - accuracy: 0.8253


[0.5913529396057129, 0.8252666592597961]

In [42]:
new_model.evaluate(X_test, y_test)  # Saved/loaded model

938/938 [==============================] - 1s 1ms/step - loss: 0.5914 - accuracy: 0.0954


[0.5913529396057129, 0.09543333202600479]

In [43]:
# Why is second number different? Because that metric is specified w/.compile!
new_model.compile(loss='sparse_categorical_crossentropy', optimizer='nadam',
                  metrics=['accuracy'])
new_model.evaluate(X_test, y_test)

938/938 [==============================] - 1s 1ms/step - loss: 0.5914 - accuracy: 0.8253


[0.5913529396057129, 0.8252666592597961]

### Stretch Goals
- Mount your Google Drive to Colab to persist your model checkpoint files. 
- Research L2 normalization (weight decay)
- Write a custom callback function to stop training after you reach .88 validation accuracy. 
- Select a new dataset and apply a neural network to it.
- Research TensorFlow Serving
- Play [QuickDraw](https://quickdraw.withgoogle.com/data)
- Create a static webpage using TensorFlow.js to serve a model. Check out [Teachable Machine Learning](https://teachablemachine.withgoogle.com/) for ideas. 